### DISCUSSION TITLE
In Stylet IoC, is it possible to bind a set of types to singleton scope automatically?

### DISCUSSION THREAD

**Discussion by mwdpb** [canton7/Stylet#340] [Q&A]
For example, my shellview managers a dozens of viewmodels which are all sub-classes of `Screen`, all those viewmodels needs to be singleton, I could explicitly configure them as 
```
builder.Bind<Page1ViewModel>().ToSelf().InSingletonScope();
builder.Bind<Page2ViewModel>().ToSelf().InSingletonScope();
...
builder.Bind<Page12ViewModel>().ToSelf().InSingletonScope();
```

It works, but is not elegant.  Is it possible to instruct IoC to use singleton scope for all classes that is derived from base class `Screen`?

Another questions is if Stylet IoC can do lazy instantiation?  Use the above example, the shellview manages a dozens of tabs (each tab corresponds to one viewmodel and only one tab will be active at a time) and some of tabs are computation intensive while some of tabs are merely user settings.  To have good user experience, I don't want all of the viewmodels being instantiated when I launch shellview since I might only need work with user settings tabs.  What's the best way to handle such situations?

**Comment by mwdpb** [comment]
For the second question, I think I got the answer by going over your excellent manual N+1 times :)

```
public interface IViewModelFactory
{
    Page1ViewModel GetPage1ViewModel();
    Page2ViewModel GetPage1ViewModel();
...
    Page12ViewModel GetPage1ViewModel();
}
```

And then pass `IViewModelFactory` as a shellviewmodel constructor parameter, I can then use the following statement in shellviewmodel to activate
```
this.ActivateItem(this.viewModelFactory.GetHomeViewModel());
```

Just want to say that the more I used your framework, the more "wow"s I shouted!  Thank you so much for your great work!

**Comment by canton7** [comment]
>  Is it possible to instruct IoC to use singleton scope for all classes that is derived from base class Screen?

Nothing built in, but it's pretty easy with reflection.

```c#
foreach (var type in someAssembly.GetTypes().Where(x => typeof(Screen).IsAssginableFrom(x)))
{
    builder.Bind(type).ToSelf().InSingletonScope();
}
```

> Just want to say that the more I used your framework, the more "wow"s I shouted! Thank you so much for your great work!

Thanks!
